Settings from file or s3

joelochlann opened the pull request:
Following on from #922 and #937

`switches` should no longer be set inside the Play config. Instead, the Play config must specify at least one of these (example values given):

```
adminSettingsSource.s3 {
  bucket="support-frontend-admin-console"
  key= "DEV/settings.json"
}
```

```
adminSettingsSource.local.path="~/.gu/support-frontend-admin-console/settings.json"
```

It will look for the `adminSettingsSource.local.path` config first. If not present, it will expect both `adminSettingsSource.s3.bucket` and `adminSettingsSource.s3.key`. If these aren't present, the app will not start up. Similarly the app will not start up if it's unable to fetch and decode the settings from the specified location.

Mullefa commented on app/admin/Settings.scala:
```diff
@@ -17,19 +24,50 @@ case class Settings(
-    )
+  def fromDiskOrS3(config: Config, s3: AmazonS3): Either[Throwable, Settings] = {
+    val adminSettingsSource: Either[Throwable, AdminSettingsSource] = if (config.hasPath("local.path")) {
+      AdminSettingsSource.fromDisk(config.getString("local.path"))
```
`getString()` can throw an exception; so the function signature of `fromDiskOrS3` is misrepresentative - we're not always encoding the error as a value

Mullefa commented on app/config/Configuration.scala:
```diff
@@ -39,6 +40,7 @@ class Configuration {
   lazy val stepFunctionArn = StateMachineArn.fromString(config.getString("supportWorkers.arn")).get
 
-  implicit val settings = Settings.fromConfig(config.getConfig("switches"))
+  private val s3 = AwsS3Client.getClient
```
My preference would be to inject the dependency on the AWS S3 client; which is marginally easier to do given `Configuration` is already a class.

Mullefa commented on app/admin/Settings.scala:
```diff
@@ -17,19 +24,50 @@ case class Settings(
-      )
-    )
+  def fromDiskOrS3(config: Config, s3: AmazonS3): Either[Throwable, Settings] = {
+    val adminSettingsSource: Either[Throwable, AdminSettingsSource] = if (config.hasPath("local.path")) {
```
Out of interest, what are your thoughts on always working with a global config vs passing around local bits of it?

joelochlann replied:
I'm happy with passing around local bits of it - but in this case it's poorly named (should be `adminSettingsSourceConfig` or similar)

joelochlann replied:
actually, thinking about this, maybe it's a bit silly to rely on the variable name to tell you which bit of the config you're in. if we always work on the global config then we don't need to worry about naming and we just assume that type `Config` means the whole config. was this what you were getting at?

Mullefa replied:
yeah, at the minor cost of  possibly longer path names, you get to reason about the config with no additional context

Mullefa commented on app/admin/Settings.scala:
```diff
@@ -17,19 +24,59 @@ case class Settings(
+  def fromDiskOrS3(config: Config)(implicit s3: AmazonS3): Either[Throwable, Settings] =
+    for {
+      source <- AdminSettingsSource.fromConfig(config)
+      rawJson <- getRawJson(source).leftMap(err => new Error(s"Could not fetch settings from $source. $err"))
```
Might be a little cleaner if the left maps happened in the respective methods. Apart from that, looks great!

joelochlann commented:
Hm, I deployed to CODE and had a failed deploy. So I've put up #945 to try and get some visibility on that. Weirdly though when I deployed master again to CODE I had another failed deploy so I'm not actually sure it's related to this PR.

Another thought: I need to update the [24/7 runbook](https://docs.google.com/document/d/1YhONdb8ZvkX5P3f8nMtkJ183pjL6785J_dppefckqzE/edit#) before deploying this because the process for disabling PayPal one-off changes with this PR (and arguably becomes trickier)

jacobwinch commented on app/admin/Settings.scala:
```diff
@@ -17,19 +24,59 @@ case class Settings(
+      source <- AdminSettingsSource.fromConfig(config)
+      rawJson <- getRawJson(source).leftMap(err => new Error(s"Could not fetch settings from $source. $err"))
+      settings <- decodeJson(rawJson).leftMap(err => new Error(s"Could not decode settings JSON from $source. $err"))
+    } yield settings
```
Might be worth logging whether we ended up reading these settings from S3 or from disk?

joelochlann commented:
> It doesn't look like we have added any new permissions to the Cloudformation template in this PR - I suspect that may be causing problems in CODE?

aha! great shout @jacobwinch. I've added logging on startup errors in #945 to help diagnose such things more easily

joelochlann commented:
> Are we planning to read these settings on some kind of schedule in the future (so that we can pick changes up without a re-deploy?)

Yes! In an upcoming PR. (Otherwise the admin console usable by non-engineers would be pretty useless!)

We might also consider a more realtime solution, posting to a notification endpoint which puts something on SQS/Kinesis to tell all the boxes to refetch from S3
